File: FishTankScreensaver/FishTankScreensaver/FishAPI.swift
import Foundation
import Cocoa

struct FishData {
    let id: String
    let imageURL: String?
    let artistName: String
    let score: Int
}

class FishAPI {
    static let backendURL = "https://fishes-be-571679687712.northamerica-northeast1.run.app"

    enum SortType: String {
        case recent = "recent"
        case popular = "popular"
        case random = "random"

        var orderBy: String {
            switch self {
            case .recent: return "CreatedAt"
            case .popular: return "score"
            case .random: return "CreatedAt"
            }
        }
    }

    /// Parse fish array manually to handle mixed types in JSON
    private static func parseFishArray(from jsonData: Data) -> [FishData] {
        guard let json = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any],
              let dataArray = json["data"] as? [[String: Any]] else {
            return []
        }

        var results: [FishData] = []
        for item in dataArray {
            guard let id = item["id"] as? String else { continue }
            let imageURL = (item["Image"] as? String) ?? (item["image"] as? String)
            let artist = (item["Artist"] as? String) ?? (item["artist"] as? String) ?? "Anonymous"
            let upvotes = (item["upvotes"] as? Int) ?? 0
            let downvotes = (item["downvotes"] as? Int) ?? 0
            let score = (item["score"] as? Int) ?? (upvotes - downvotes)

            results.append(FishData(id: id, imageURL: imageURL, artistName: artist, score: score))
        }
        return results
    }

    static func fetchFish(sort: SortType, limit: Int = 50, completion: @escaping ([FishData]) -> Void) {
        var components = URLComponents(string: "\(backendURL)/api/fish")!
        var queryItems = [
            URLQueryItem(name: "limit", value: "\(limit)"),
            URLQueryItem(name: "order", value: "desc"),
            URLQueryItem(name: "isVisible", value: "true"),
            URLQueryItem(name: "deleted", value: "false"),
            URLQueryItem(name: "orderBy", value: sort.orderBy)
        ]
        if sort == .random {
            queryItems.append(URLQueryItem(name: "random", value: "true"))
        }
        components.queryItems = queryItems

        guard let url = components.url else {
            FishLog.log("fetchFish: invalid URL")
            DispatchQueue.main.async { completion([]) }
            return
        }

        FishLog.log("fetchFish: requesting \(url.absoluteString)")

        let task = URLSession.shared.dataTask(with: url) { data, response, error in
            if let error = error {
                FishLog.log("fetchFish: network error: \(error.localizedDescription)")
                DispatchQueue.main.async { completion([]) }
                return
            }
            guard let data = data else {
                FishLog.log("fetchFish: no data returned")
                DispatchQueue.main.async { completion([]) }
                return
            }
            if let httpResponse = response as? HTTPURLResponse {
                FishLog.log("fetchFish: HTTP \(httpResponse.statusCode), \(data.count) bytes")
            }

            let results = parseFishArray(from: data)
            FishLog.log("fetchFish: parsed \(results.count) fish")
            DispatchQueue.main.async { completion(results) }
        }
        task.resume()
    }

    /// Downloads image data on a background thread
    static func loadImageData(from urlString: String, completion: @escaping (Data?) -> Void) {
        guard let url = URL(string: urlString) else {
            completion(nil)
            return
        }
        let task = URLSession.shared.dataTask(with: url) { data, _, error in
            if let error = error {
                FishLog.log("loadImage: error: \(error.localizedDescription)")
            }
            completion(data)
        }
        task.resume()
    }
}

// Simple file-based logger for debugging screensaver issues
class FishLog {
    static let logPath = "/tmp/fishtank_screensaver.log"

    static func log(_ message: String) {
        let timestamp = ISO8601DateFormatter().string(from: Date())
        let line = "[\(timestamp)] \(message)\n"
        if let handle = FileHandle(forWritingAtPath: logPath) {
            handle.seekToEndOfFile()
            handle.write(line.data(using: .utf8)!)
            handle.closeFile()
        } else {
            FileManager.default.createFile(atPath: logPath, contents: line.data(using: .utf8))
        }
    }
}

File: FishTankScreensaverWindows/FishAPI.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;

namespace FishTankScreensaver
{
    public class FishData
    {
        public string Id { get; set; } = "";
        public string? ImageURL { get; set; }
        public string ArtistName { get; set; } = "Anonymous";
        public int Score { get; set; }
    }

    public static class FishAPI
    {
        private static readonly string BackendURL = "https://fishes-be-571679687712.northamerica-northeast1.run.app";
        private static readonly HttpClient Http = new();

        public static string GetOrderBy(string sortType)
        {
            return sortType switch
            {
                "popular" => "score",
                _ => "CreatedAt"
            };
        }

        public static async Task<List<FishData>> FetchFishAsync(string sortType, int limit = 50)
        {
            var results = new List<FishData>();
            try
            {
                var orderBy = GetOrderBy(sortType);
                var url = $"{BackendURL}/api/fish?limit={limit}&order=desc&isVisible=true&deleted=false&orderBy={orderBy}";
                if (sortType == "random")
                    url += "&random=true";

                FishLog.Log($"FetchFish: requesting {url}");

                var response = await Http.GetStringAsync(url);
                FishLog.Log($"FetchFish: received {response.Length} chars");

                using var doc = JsonDocument.Parse(response);

                if (!doc.RootElement.TryGetProperty("data", out var dataArray))
                {
                    FishLog.Log("FetchFish: no 'data' property in response");
                    return results;
                }

                foreach (var item in dataArray.EnumerateArray())
                {
                    var id = item.TryGetProperty("id", out var idProp) ? idProp.GetString() : null;
                    if (id == null) continue;

                    string? imageUrl = null;
                    if (item.TryGetProperty("Image", out var img))
                        imageUrl = img.GetString();
                    else if (item.TryGetProperty("image", out var img2))
                        imageUrl = img2.GetString();

                    string artist = "Anonymous";
                    if (item.TryGetProperty("Artist", out var art))
                        artist = art.GetString() ?? "Anonymous";
                    else if (item.TryGetProperty("artist", out var art2))
                        artist = art2.GetString() ?? "Anonymous";

                    int score = 0;
                    if (item.TryGetProperty("score", out var sc) && sc.ValueKind == JsonValueKind.Number)
                        score = sc.GetInt32();
                    else
                    {
                        int up = 0, down = 0;
                        if (item.TryGetProperty("upvotes", out var upProp) && upProp.ValueKind == JsonValueKind.Number)
                            up = upProp.GetInt32();
                        if (item.TryGetProperty("downvotes", out var downProp) && downProp.ValueKind == JsonValueKind.Number)
                            down = downProp.GetInt32();
                        score = up - down;
                    }

                    results.Add(new FishData
                    {
                        Id = id,
                        ImageURL = imageUrl,
                        ArtistName = artist,
                        Score = score
                    });
                }

                FishLog.Log($"FetchFish: parsed {results.Count} fish");
            }
            catch (Exception ex)
            {
                FishLog.Log($"FetchFish ERROR: {ex}");
            }
            return results;
        }

        public static async Task<byte[]?> LoadImageDataAsync(string url)
        {
            try
            {
                return await Http.GetByteArrayAsync(url);
            }
            catch (Exception ex)
            {
                FishLog.Log($"LoadImage ERROR ({url}): {ex.Message}");
                return null;
            }
        }
    }
}
